src/decorators/parse-required-and-nullable-validation.helper.ts

```diff
@@ -0,0 +1,52 @@
+import { ValidateIf } from 'class-validator';
+
+export type RequiredOption = OptionalBooleanOrBooleanDependingOnCurrentObject;
+export type NullableOption = OptionalBooleanOrBooleanDependingOnCurrentObject;
+
+type OptionalBooleanOrBooleanDependingOnCurrentObject = undefined | BooleanOrBooleanDependingOnCurrentObject;
+type BooleanOrBooleanDependingOnCurrentObject = boolean | BooleanDependingOnCurrentObject;
+type BooleanDependingOnCurrentObject = (currentObject: Record<string, unknown>) => boolean;
+export const parseRequiredAndNullable = ({ required, nullable }: RequiredAndNullable): ParsedRequiredAndNullableOptions => {
+  const requiredOrDefault = required ?? true;
+  const nullableOrDefault = nullable ?? false;
+  return {
+    shouldPropertyBeDocumentedAsRequired: typeof requiredOrDefault === 'function' ? false : requiredOrDefault,
+    shouldPropertyBeDocumentedAsNullable: typeof nullableOrDefault === 'function' ? true : nullableOrDefault,
+    validationDecoratorsToApplyForRequiredOption: [
+      ValidateIf(
+        (currentObject, propertyValue) =>
+          !getAllowedNullishValuesForProperty(currentObject, { required: requiredOrDefault, nullable: nullableOrDefault }).includes(propertyValue),
+      ),
+    ],
+  };
+};
+
+const asFunctionDependingOnCurrentObject = (x: BooleanOrBooleanDependingOnCurrentObject): BooleanDependingOnCurrentObject =>
+  typeof x === 'function' ? x : () => x;
+
+const getAllowedNullishValuesForProperty = (currentObject: Record<string, unknown>, { required, nullable }: RequiredAndNullable): (undefined | null)[] => {
+  const propertyIsRequiredOnCurrentObject = asFunctionDependingOnCurrentObject(required);
+  const propertyIsNullableOnCurrentObject = asFunctionDependingOnCurrentObject(nullable);
+  const allowedNullishValues: (undefined | null)[] = [];
+
+  if (propertyIsNullableOnCurrentObject(currentObject)) {
+    allowedNullishValues.push(null);
+  }
+
+  if (!propertyIsRequiredOnCurrentObject(currentObject)) {
+    allowedNullishValues.push(undefined);
+  }
+
+  return allowedNullishValues;
+};
+
+interface RequiredAndNullable<> {
+  required: RequiredOption;
+  nullable: NullableOption;
+}
+
+interface ParsedRequiredAndNullableOptions {
+  shouldPropertyBeDocumentedAsRequired: boolean;
+  shouldPropertyBeDocumentedAsNullable: boolean;
+  validationDecoratorsToApplyForRequiredOption: PropertyDecorator[];
```

Comment:
Should this be `validationDecoratorsToApplyForRequiredAndNullableOptions`?

Reply:
I've gone for `validationDecoratorsToApply`